packages/block-library/src/navigation-link/hooks.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import {
+	category,
+	page,
+	postTitle,
+	tag,
+	customPostType,
+} from '@wordpress/icons';
+
+function getIcon( variationName ) {
+	switch ( variationName ) {
+		case 'post':
+			return postTitle;
+		case 'page':
+			return page;
+		case 'tag':
+			return tag;
+		case 'category':
+			return category;
+		default:
+			return customPostType;
+	}
+}
+
+function enhanceNavigationLinkVariations( settings, name ) {
+	if ( name !== 'core/navigation-link' ) {
+		return settings;
+	}
+
+	if ( settings?.variations ) {
+		const variations = settings.variations.map( ( variation ) => {
+			return {
+				...variation,
+				...( ! variation.icon && {
+					icon: getIcon( variation.name ),
+				} ),
+				...( ! variation.isActive && {
+					isActive: ( blockAttributes, variationAttributes ) => {
+						return (
+							blockAttributes.type === variationAttributes.type
+						);
+					},
+				} ),
+			};
+		} );
+		return {
+			...settings,
+			variations,
+		};
+	}
+	return settings;
+}
+
+addFilter(
```

Comment:
The fact that we register this filter as a side effect next to its definition isn't ideal but I guess it's fine since it's a special case.

Reply:
I think it's inevitable to use the filter on the client. I raised it mostly so you were aware of the implications it creates.

Reply:
@gziolo Do folks mind if we keep this as is for now? The filter does run if we exclude the navigation link from the block library loader, but the alternative of adding it in __experimentalRegisterExperimentalCoreBlocks gets relatively messy.